---
title: exercise-r-main-lzl
output:
  html_document:
    df_print: paged
  word_document: default
  pdf_document: default
---

# Programming Exercises for R

```{r}
library(readr)
library(ggplot2)
library(dplyr)
library(tidyverse)
library(data.table)
```

# 
# Mutation data analysis 

1. Identify the top 10 most frequently mutated genes. 
   Identify samples whose mutation count is in the 80 to 90 percentile.
```{r}
mut_data <- readRDS("data/mutations_sclc_ucologne_2015.rds")

# Identify the top 10 most frequently mutated genes.
gene_mut <- mut_data %>%
  group_by(gene) %>%
  summarise(mutation_num=n()) %>%
  arrange(-mutation_num)

top_10_fre_mut_gene <- head(gene_mut,11)$gene
top_10_fre_mut_gene
# The top 10 most frequently mutated genes are:
# TTN, TP53,MUC16, RB1, RYR2, LRP1B, ZFHX4, CSMD3, USH2A, NAV3，SYNE1

# Identify samples whose mutation count is in the 80 to 90 percentile
sample_mut <-  mut_data %>%
  group_by(sample_id) %>%
  summarise(mutation_num=n()) %>%
  arrange(-mutation_num)


sample_fitered <- sample_mut %>%
  filter(mutation_num >= quantile(sample_mut$mutation_num,0.8) & mutation_num <= quantile(sample_mut$mutation_num,0.9))

sample_fitered$sample_id 

```



2. Categorize variants based on their expected effects using the`data/mutation_effects.tsv` table.
Generate a count matrix containing the numbers of loss-of-function and neutral mutations for each gene. 
```{r}
mut_effect <- readr::read_tsv("data/mutation_effects.tsv")

# Categorize variants based on their expected effects
va_category <- list()
for (i in unique(mut_effect$effect)) {
va_category[[i]] <- (mut_effect[mut_effect$effect==i,])$variant_class
}

# Generate a count matrix containing the numbers of loss-of-function and neutral mutations for each gene
matrix_1 <- mut_data %>%
  filter(variant_class %in% va_category[["loss_of_function"]]) %>%
  group_by(gene) %>%
  summarise(num_loss_of_function = n())
  
matrix_2 <- mut_data %>%
  filter(variant_class %in% va_category[["neutral"]]) %>%
  group_by(gene) %>%
  summarise(num_neural = n())

count_matrix <- full_join(matrix_1, matrix_2, by = "gene")
count_matrix[is.na(count_matrix)] <- 0
count_matrix

rm(matrix_1, matrix_2)

```


3. Implement a statistical test that determines whether a gene has a significantly higher proportion of loss-of-function mutations (excluding mutations with uncertain effects),compared to other genes.

```{r}
df <- count_matrix %>%
  mutate(other_num_loss_of_function=sum(count_matrix$num_loss_of_function)-num_loss_of_function) %>%
  mutate(other_num_neural=sum(count_matrix$num_neural)-num_neural)

# Define a function to perform Fisher's exact test and calculate odds ratio
perform_test <- function(x) { 
  p <- fisher.test(x, alternative="greater")$p.value
  odds_ratio <- (x[1,1]*x[2,2])/(x[1,2]*x[2,1])
  return(c("p.value" = p, "estimated_effect_size" = odds_ratio))
}


result <- t(apply(df, 1, 
      function(x) {
          perform_test(matrix(as.numeric(x[2:5]), ncol=2, byrow=T))
      }))

df <- cbind(df, result)
# FDR correction
df$q.value <- p.adjust(df$p.value,method = "BH")

```


4. Identify candidate tumour suppressor genes using this statistical test,
   adjusting for multiple hypothesis testing.
   The output table should contain:
     - gene symbol
     - an estimate of effect size
     - p value
     - q value

```{r}
da <- df %>%
  dplyr::select(gene, estimated_effect_size, p.value, q.value)

candidate_gene <- da %>% filter(q.value < 0.05)
candidate_gene
```


 5. Perform a literature search and explain the function of each candidate
   gene in the context of cancer, as well as specifically in small cell 
   lung cancer.

MLL2:
MLL2 encodes a histone methyltransferase, a key regulator of transcriptional enhancer function. MLL2 mutations occur in several cancer types such as breast cancer and colon cancer. MLL2 is frequently mutated in SCLC.
MLL2 loss is an important driver of SCLC. MLL2 mutation in human SCLC cell lines was associated with reduced lysine methyltransferase 2D protein levels and reduced mono-methylation of histone H3 lysine 4 (H3K4me1), a chromatin marker associated with transcriptional enhancers. The functional role of MLL2 in SCLC tumorigenesis and tumor cell plasticity remains unexplored.

RB1:
RB1 is a critical negative regulator of cell cycle progression through its ability to repress E2F target genes. It serves as a tumor suppressor gene that is inactivated in many tumor types and is only frequently mutated in SCLC amongst adult cancers. SCLC is almost linked to inactivating RB1 and TP53 mutations.
The Rb protein encoded by RB1 gene is a key component of the p16ink4A –Cdk/cyclin–Rb pathway responsible for cell division arrest at the G1/S checkpoint.The lack of Rb activity correlated with lower survival of patients with various types of lung cancer. 

TP53:
The TP53 protein exerts its tumor-suppressor function as a sequence-specific transcription factor.The transcription factor TP53 can activate the transcription of numerous downstream genes, such as p21 and MDM2, by binding to specific sequences, which often mediates their biological functions. Under normal conditions, TP53 is rapidly degraded and not present at detectable levels within the cell. Various types of cellular stress result in the stabilization and activation of TP53, causing protein accumulation within the nucleus. The TP53 pathway is activated by such cellular stresses that alter the normal cell-cycle progression or can induce mutations of the genome, leading to the transformation of a normal cell into a cancerous cell. The TP53 protein is a negative regulator of cell proliferation and a positive regulator of apoptosis in response to DNA damaging agents. 
TP53 is highly mutated in many types of cancers and appears to be the most frequent target. Abnormality of TP53 plays an important role in the tumorigenesis of lung epithelial cells and most SCLC patients have alterations in this gene.



# Transcriptomic data normalization

The transcriptomic data is in `data/expr_sclc_ucologne_2015.rds`.

1. Perform an appropriate log transformation on the data.
```{r}
exp_data <- readRDS("data/expr_sclc_ucologne_2015.rds")
log_exp_data <- log2(exp_data + 1)

```

2. Implement a median polish algorithm from scratch.
```{r}
medpol <- function(x) {
  r_effect <- rep(0,nrow(x))
  c_effect <- rep(0,ncol(x))
  global_effect <- 0
  eps <- 0.01
  max_iteration <- 10
  oldsum <- 0
  Residual <- x  
  
  for (i in range(max_iteration)) {
    # Apply polishing by row median
    Rmed <- apply(Residual, 1, median)
    Residual <- sweep(Residual, 1, Rmed)
    r_effect <- r_effect + Rmed
    delta <- median(c_effect)
    c_effect <- c_effect - delta
    global_effect <- global_effect + delta
    
    # Apply polishing by column median
    Cmed <- apply(Residual, 2, median)
    Residual <- sweep(Residual, 2, Cmed)
    c_effect <- c_effect + Cmed
    delta <- median(r_effect)
    r_effect <- r_effect - delta
    global_effect <- global_effect + delta    
    
    newsum <- sum(abs(Residual))
    # Convergence check
    if (newsum == 0 || abs(newsum-oldsum) < eps*newsum) {
      break
    }
    oldsum <- newsum
  }
  if (!(newsum == 0 || abs(newsum-oldsum) < eps*newsum)) {
    cat("Warning: Polish did not converge in",max_iteration,"iterations")
    
    }
  list(Residual=Residual, row_effect=r_effect, col_effect=c_effect, global_effect=global_effect)
}

# Implement the median polish algorithm
my_data_polish <- medpol(log_exp_data)

```

3. Compare the residuals of your algorithm and `stats::medpolish`.
```{r}
# Residual of my algorithm
my_residual <- my_data_polish$Residual

# Residual of `stats::medpolish`
data_polish2 <- stats::medpolish(log_exp_data)
residual2 <- data_polish2$residuals 

# Comparison
all.equal(my_residual,residual2)
```

4. Plot heatmaps of the results before and after median polish.
```{r}
library(gplots)
library(RColorBrewer)

hmcol <- rev(brewer.pal(11, "RdBu"))
#  heatmaps of the results before median polish
heatmap.2(log_exp_data, 
                col=hmcol, # define the customized colormap
                scale="none", 
                density.info="none",
                trace="none",
                Colv=F,
                cexRow = 1
)
#  heatmaps of the results after median polish
heatmap.2(my_residual, 
                col=hmcol, 
                scale="none", 
                density.info="none", 
                trace="none",
                Colv=F,
                cexRow = 1
)



```

5. Output the median polished residual matrix as the normalized transcriptomic data.
```{r}
transcript_data <- my_residual
write.csv(transcript_data,"normalized_transcriptomics_data.csv")
```


# Differential gene expression analysis

The clinical data is in `data/pheno_sclc_ucologne_2015.tsv`.

1. Define two groups of tumours as early stage (stages I-II) vs. advanced stage
   tumours (stages III-IV), while excluding samples missing stage information.
   
```{r}
clinical_data <- readr::read_tsv("data/pheno_sclc_ucologne_2015.tsv")

clinical <- clinical_data %>%
  mutate(group=ifelse(str_detect(uicc_tumor_stage, "^(I|II)[abAB]*$"), "early_stage", ifelse(str_detect(uicc_tumor_stage,"^(III|IV)[abAB]*$"),"advanced_stage", NA))) %>%
dplyr::filter(!is.na(group)) %>%
dplyr::select(patient_id, group)

clinical

```  
   
2. Identify genes that differentially expressed in early vs. advanced stage
   tumours using an appropriate R package.

```{r}
# Differentially expression analysis by limma
library(limma)

# Filter transcript data in samples in early or advanced stage
samples <- as.vector(intersect(colnames(transcript_data), clinical$patient_id))
transcript_data_samples <- transcript_data[,samples]
# ncol(dat)
# nrow(dat)

# Filter clinical data to get group information 
# Sort the group information in order
clinical_of_samples <- clinical %>% 
  dplyr::filter(patient_id %in% samples)
# ncol(group)
# nrow(group)

# Sort the columns of gene expression matrix based on the order of patient_ids in the group information
transcript_data_samples_sorted <- transcript_data_samples[,match(transcript_data_samples_sorted$patient_id,colnames(dat))]
all(clinical_of_samples$patient_id==colnames(transcript_data_samples_sorted))

group_factor <- factor(clinical_of_samples$group, levels=c("early_stage","advanced_stage"),ordered = F)

# Experimental group: advanced_stage   Control group: early_stage
design <-model.matrix(~group_factor)
fit<-lmFit(dat_sorted,design)
fit<-eBayes(fit)
options(digits = 4)
deg<-topTable(fit,adjust='BH',number = Inf)

# Get the up, down, stable genes
cut_off_pvalue <- 0.05
cut_off_logFC <- 1

deg$change <- ifelse(deg$P.Value < cut_off_pvalue & abs(deg$logFC) >= cut_off_logFC, 
                    ifelse(deg$logFC>= cut_off_logFC,'Up','Down'),'Stable')

# Identify the up and down genes as DEGs
DEGs <- deg[deg$change!="Stable",]

DEGs
```

# Integrative analysis

The structural variant data is in `data/sv_sclc_ucologne_2015.tsv`.

1. For each gene involved in a structural variant (SV), determine the expression 
   level of the gene in the sample that harbours the SV.
```{r}

sv_data <- readr::read_tsv("data/sv_sclc_ucologne_2015.tsv")

sv_data_exp <- sv_data %>%
  rowwise() %>%
  mutate(site1_hugo_symbol_exp = ifelse(site1_hugo_symbol %in% rownames(transcript_data), 
                                        transcript_data[site1_hugo_symbol, sample_id], 
                                        NA)) %>%
  mutate(site2_hugo_symbol_exp = ifelse(site2_hugo_symbol %in% rownames(transcript_data), 
                                        transcript_data[site2_hugo_symbol, sample_id], 
                                        NA))  
sv_data_exp 
```
   
2. Identify SVs that satisfy the following criteria:
      - The involved pair of genes both have elevated expression levels in samples with the SV compared to samples without the SV.
      - The second gene in the pair is in frame.

```{r}
# function to get log2fc
diff_exp <- function(transcript_data, gene, sample_id) {
  case_exp <- transcript_data[gene,sample_id]
  control_exp <- mean(transcript_data[gene,-which(colnames(transcript_data)==sample_id)])
  log2fc <- case_exp-control_exp
  return(log2fc)
}

threshould <- 0
# Identify the involved pair of genes both have elevated expression levels and the second gene in the pair is in frame
sv_identified <- sv_data %>%
  dplyr::filter(site1_hugo_symbol %in% rownames(transcript_data), 
                site2_hugo_symbol %in% rownames(transcript_data),
                site2_effect_on_frame == "in-frame") %>%
  rowwise() %>%
  dplyr::filter(diff_exp(transcript_data, site1_hugo_symbol, sample_id) > threshould, 
         diff_exp(transcript_data, site2_hugo_symbol, sample_id) > threshould) 


sv_identified
```



